### PR TITLE
Issue 187 - Explain API

### PR DESCRIPTION
**Pull request by worleydl**
Initial changes to add the explain API.

TODO:

- [X] Build out tests
- [X] Move SnapshotRewriter code to querqy-core

### CODE REVIEW

**Comment by epugh** [comment]
Does this all get dumped out in the Solr debug or is this a new thing?    I'd love to expose this in Quepid...

**Comment by renekrie** [comment]
@epugh It's a new HTTP endpoint: #187

File: querqy-for-lucene/querqy-solr/src/main/java/querqy/solr/explain/SnapshotRewriter.java [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+        @Override
+        public Void visit(final Term term) {
+            final Map<String, Object> tq = new HashMap<>();
+            final HashMap props = new HashMap<>();
+
+            if (term.getField() != null) {
+                props.put("field", term.getField());
+            }
+
+            props.put("term", term.toString());
```
Comment (renekrie) [inline querqy-for-lucene/querqy-solr/src/main/java/querqy/solr/explain/SnapshotRewriter.java:204]: `term` and `generated' could be constants

File: querqy-for-lucene/querqy-solr/src/main/java/querqy/solr/explain/SnapshotRewriter.java [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+            super.visit(disjunctionMaxQuery);
+            clauses = origClauses;
+
+            return null;
+
+        }
+
+        @Override
+        public Void visit(final Term term) {
```
Comment (renekrie) [inline querqy-for-lucene/querqy-solr/src/main/java/querqy/solr/explain/SnapshotRewriter.java:196]: `term` could also be of type `BoostedTerm` in that case we should add the boost to the snapshot

File: querqy-for-lucene/querqy-solr/src/main/java/querqy/solr/explain/SnapshotRewriter.java [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+            final Map<String, Object> tq = new HashMap<>();
+            final HashMap props = new HashMap<>();
+
+            if (term.getField() != null) {
+                props.put("field", term.getField());
```
Comment (renekrie) [inline querqy-for-lucene/querqy-solr/src/main/java/querqy/solr/explain/SnapshotRewriter.java:201]: make 'field' a constant

File: querqy-for-lucene/querqy-solr/src/main/java/querqy/solr/explain/SnapshotRewriter.java (hunk 1 above)
Comment (renekrie) [reply querqy-for-lucene/querqy-solr/src/main/java/querqy/solr/explain/SnapshotRewriter.java:204]: I think we should rather use " props.put("value", term.toString());"

**Review by renekrie — CHANGES_REQUESTED** [review]
ExplainRewriter: Please deal with BoostedTerm and with Term.value property name. Nice to have: use constants where possible

File: querqy-for-lucene/querqy-solr/src/main/java/querqy/solr/explain/SnapshotRewriter.java [hunk 4]
```diff
@@ -0,0 +1,247 @@
+package querqy.solr.explain;
+
+import org.apache.solr.common.util.NamedList;
+import org.graalvm.compiler.lir.phases.PreAllocationOptimizationPhase;
```
Comment (mkr) [inline querqy-for-lucene/querqy-solr/src/main/java/querqy/solr/explain/SnapshotRewriter.java:4]: Unused import ;)

**Review by renekrie — APPROVED** [review]

**Review by renekrie — APPROVED** [review]

**Comment by renekrie** [comment]
Thank you @worleydl !

**Pull request merged**